### ISSUE TITLE
H1 zero-copy issue with client introduced delays between header and chunked data

### ISSUE THREAD

**Issue by awlx** [haproxy/haproxy#2598]
### Detailed Description of the Problem

When a client introduces roughly a 200ms delay between sending the header and the payload chunked data, haproxy truncates the first 10 bytes of the payload when sending it to the backend.

I sniffed the data with an eBPF tracer after the SSL decrypt and this is what we see:
```
 t=2128741385690363 type=read data=POST /path HTTP/1.1
Authorization: token
Accept: application/vnd.app.v3+json
User-Agent: User_agent/2.18
Transfer-Encoding: chunked
Host: localhost
Connection: Keep-Alive
Accept-Encoding: gzip,deflate

1f3e

```

You can see 1f3e which is the correct content length of 7998 bytes (I truncated the tail of the of the payload as it doesn't matter).  Then the client begins to introduce a significant delay but haproxy already starts sending the headers to the backend (which are complete and correct).

After some time (~200ms) the chunked data arrives from the client
```
t=2127677488967861 type=read data={"ref":"refs/heads/main","sarif":"H4sIAAAAAAAA
```

And haproxy passes this mangled thing on to the server:
```
t=2127677489108060 type=write data=00001F3E
fs/heads/main","sarif":"H4sIAAAAAAAA
```

As you can see 00001F3E and the `\r\n` got mixed into this message (which are the 10bytes) and the beginning of the json is missing. This ultimately causes a 400 bad request on the server end as the json is broken.

The current workaround is to disable zero-copy like this `tune.h1.zero-copy-fwd-send off` but this doesn't seem like intended behaviour.

### Expected Behavior

Don't mix the payload with the header and don't truncate the payload by 10 bytes.

### Steps to Reproduce the Behavior

For Python

```
REQUEST = b'Headers'
CHUNK = b'chunkeddata'
with socket.create_connection((ip, 443)) as sock:
    with context.wrap_socket(sock, server_hostname='localhost') as ssock:
        ssock.send(REQUEST)
        time.sleep(0.3)
        ssock.send(CHUNK)
        while True:
         data = ssock.recv()
         if not data:
            break
         print(data)
```

### Do you have any idea what may have caused this?

To me it sounds a bit like haproxy is entering the fast-forward mode too early and doesn't wait for the chunks to arrive in time and then a pointer is not aligning anymore.

### What is your configuration?

global
  nbthread 8
  tune.listener.default-shards by-thread
  tune.bufsize 68608
  tune.maxrewrite 4096
  tune.pipesize 262144
  tune.ssl.maxrecord 1370
  tune.ssl.cachesize 0
  tune.ssl.capture-buffer-size 2048
  tune.h2.initial-window-size 67108864
  tune.stick-counters 15 # default is 3
  spread-checks 3

defaults https
  mode http
  balance leastconn
  option redispatch 1

  option forwardfor
  option http-server-close

    # logging
  log global

  no option dontlognull
  option log-health-checks
  option log-separate-errors
  timeout client 10s
  timeout connect 1s
  timeout server 11s
  timeout queue 15s
  timeout tunnel 30s
  timeout http-keep-alive 30s
  timeout http-request 5s
  timeout tarpit 30s
  timeout client-fin 2s
  timeout server-fin 5s

### Output of `haproxy -vv`

HAProxy version 3.0.0 2024/05/29 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2029.
Known bugs: http://www.haproxy.org/bugs/bugs-3.0.0.html
Running on: Linux 5.15.0-105-generic #115~20.04.1-Ubuntu SMP Mon Apr 15 17:33:04 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fwrapv -D_FORTIFY_SOURCE=2 -fstack-protector-strong -fPIC -fwrapv -DMAX_SYSLOG_LEN=2048
  OPTIONS = USE_THREAD=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_ZLIB=1 USE_TFO=1 USE_SYSTEMD=1 USE_STATIC_PCRE2=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   =

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY -LUA -MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL -PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN -SLZ +SSL -STATIC_PCRE +STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=16).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020 
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020 
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.
Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX 
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace

**Comment by capflam** [comment]
I can confirm the issue. I'm on it. Thanks !

**Comment by capflam** [comment]
It should be fixed now. A 3.0.1 will be emitted this afternoon shipping the fix. Thanks !

**Comment by awlx** [comment]
Thanks so much for this quick fix @capflam ! I've run our tests against 3.0.1 and can't reproduce the issue anymore. 🥳
